src/lib.rs

```diff
@@ -162,3 +219,206 @@ impl LevelIndex {
+        let model = shift_and_mask_lower(0, 8, v);
+        let primaries = shift_and_mask_lower(8, 8, v);
+        let transfer = shift_and_mask_lower(16, 8, v);
+        let flags = shift_and_mask_lower(24, 8, v);
+
+        let v = bytes_to_u32(bytes, &mut offset);
+        let texel_block_dimensions = [
+            shift_and_mask_lower(0, 8, v) + 1,
+            shift_and_mask_lower(8, 8, v) + 1,
+            shift_and_mask_lower(16, 8, v) + 1,
+            shift_and_mask_lower(24, 8, v) + 1,
+        ];
+
+        let v = bytes_to_u32(bytes, &mut offset);
+        let mut bytes_planes = [0u32; 8];
+        bytes_planes[0] = shift_and_mask_lower(0, 8, v);
+        bytes_planes[1] = shift_and_mask_lower(8, 8, v);
+        bytes_planes[2] = shift_and_mask_lower(16, 8, v);
+        bytes_planes[3] = shift_and_mask_lower(24, 8, v);
+
+        let v = bytes_to_u32(bytes, &mut offset);
+        bytes_planes[4] = shift_and_mask_lower(0, 8, v);
+        bytes_planes[5] = shift_and_mask_lower(8, 8, v);
+        bytes_planes[6] = shift_and_mask_lower(16, 8, v);
+        bytes_planes[7] = shift_and_mask_lower(24, 8, v);
+
+        Self {
+            vendor_id,
+            descriptor_type,
+            version_number,
+            descriptor_block_size,
+            color_model: ColorModel::new(model),
+            color_primaries: ColorPrimaries::new(primaries),
+            transfer_function: TransferFunction::new(transfer),
+            flags: DataFormatFlags::from_bits_truncate(flags),
+            texel_block_dimensions,
+            bytes_planes,
+            sample_data: &bytes[offset..initial_offset + descriptor_block_size as usize],
+        }
+    }
+
+    pub fn sample_information(&self) -> impl Iterator<Item = SampleInformation> + 'data {
+        SampleInformationIterator {
+            data: self.sample_data,
+            offset: 0,
+        }
+    }
+}
+
+struct SampleInformationIterator<'data> {
+    data: &'data [u8],
+    offset: usize,
+}
+
+impl<'data> Iterator for SampleInformationIterator<'data> {
+    type Item = SampleInformation;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.offset > self.data.len() - SampleInformation::LENGTH {
+            return None;
+        }
+        let sample_information =
+            SampleInformation::from_bytes(&self.data[self.offset..self.offset + SampleInformation::LENGTH]);
+        self.offset += SampleInformation::LENGTH;
+        Some(sample_information)
+    }
+}
+
+#[derive(Debug)]
+pub struct SampleInformation {
+    pub bit_offset: u32,                                //: 16;
+    pub bit_length: u32,                                //: 8;
+    pub channel_type: u32,                              //: 4;
+    pub channel_type_qualifiers: ChannelTypeQualifiers, //: 4;
+    pub sample_positions: [u32; 4],                     //: 8 x 4;
+    pub lower: u32,                                     //;
+    pub upper: u32,                                     //;
```

Comment:
From the use in `is_norm` I gather these values have structure we're not capturing. Is there a way we could abstract over them more ergonomically?

Reply:
The is_norm is my own invention based on what I saw in files. The bounds can be anything anyone wants, but they will commonly be signed/unsigned int / unorm/snorm/float. But this is kind of specific I guess so probably better to keep this code outside the crate where the usage of it is known.

Reply:
Ah, I had guessed these were "lower/upper 32 bits", not "lower/upper bound". That makes more sense. A glance at the spec suggests that the additional structure here is quite complex so passing through the bits as done here is probably the right place for us to start.